docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+
+This is the most intricate and potentially time-consuming part of the upgrade. Basically, it involves going through your entire codebase to remove references to deprecated APIs and update your project logic.
+
+### Automatically update deprecated API references with the `inspect` command
+
+The upgrader has an `inspect` command that can flag deprecated API usage, and in some cases, update your codebase to the SilverStripe 4 equivalent. This does require you to carefully review each change and warning to make sure the updated logic still work as intended. Even so, it is a huge time-saver compared to reviewing your code base manually.
+
+Note that the `inspect` command loads your files with PHP interpreter. So basic syntax errors — for example, extending a class that does not exists — will cause an immediate failure. For this reason, you need to complete [Step 4 - Update codebase with references to newly namespaced classes](#step4) before running the `inspect` command.
+
+```bash
+/var/www/SS_acme > upgrade-code inspect ./mysite/ --write
+```
+
+You can omit the `--write` flag if you just want to view the proposed changes without applying them. You can run the command on a specific subdirectory or file. This can be more manageable if you have a big project to go through.
+
+Like the `upgrade` command, `inspect` gets its list of API changes from `.upgrade.yml` files. So you may get upgrade suggestions and substitution from third party modules. You can even include your own project specific changes in your `.upgrade.yml` if you want.
+
+#### Sample output of the `inspect` command
+Here's some sample output of what you might get back the `inspect` command.
+
+```bash
+/var/www/SS_acme > upgrade-code inspect ./mysite/Models/Coyote.php
+Running post-upgrade on "/var/www/SS_acme/mysite/code/Models/Coyote.php"
+[2018-06-06 13:35:38] Applying ApiChangeWarningsRule to Coyote.php...
+modified:	Coyote.php
+@@ -68,7 +68,7 @@
+     {
+         // Getting a reference to Coyote's list of crazy ideas
+-        $manyManyRelation = $this->manyManyComponent('CrazyIdeas');
++        $manyManyRelation = $this->getSchema()->manyManyComponent('CrazyIdeas');
+         return $manyManyRelation;
+     }
+
+Warnings for Coyote.php:
+ - Coyote.php:20 SS_Cache: Using symfony/cache now (https://docs.silverstripe.org/en/4/changelogs/4.0.0#cache)
+ - Coyote.php:42 SilverStripe\Control\Director::setUrlParams(): Method removed
+ - Coyote.php:71 SilverStripe\ORM\DataObject->manyManyComponent(): DataObject->manyManyComponent() moved to DataObjectSchema. Access through getSchema(). You must manually add static::class as the first argument to manyManyComponent()
+Changes not saved; Run with --write to commit to disk
+```
+
+### Manually update deprecated API references
+
+SilverStripe 4 introduces thousands of small and big API changes. To update deprecated API references manually, you have to go through each one of your project files.
```

Comment:
Just say "many" rather than "thousands". Remember, the aim here is to make decision makers *more* confident in upgrades, not *less* confident.

Reply:
Done